Automattic/woocommerce-services
# Catch exception if the "DHL live rates available" note cannot be created

## Description


It is possible for WC Admin to be available but disabled if the `woocommerce_admin_disabled` filter is used. If this happens,
attempts to create a note will fail due to the "admin-note" data store being unavailable.

This commit adds a check if an exception is thrown on note creation and removal attempts.

### Some more details on the issue

I researched the way WC Admin is initialized to learn more about the context of how we are adding the note and here is what I learned:

We are adding the note during the `woocommerce_init` hook, which is called during the `init` hook, which is fired *after* `plugins_loaded` - meaning all plugins are loaded at this time. So far, so good.

The problem seems to be that when [the `woocommerce_admin_disabled` filter](https://github.com/woocommerce/woocommerce-admin/blob/0b68c1d7c0c5025789071022efa129126da1ded5/src/FeaturePlugin.php#L67) is set to `true`, even though WC Admin is loaded and all of its classes and traits are available (which is what we currently check for), it hasn't really registered its "admin-note" data store yet, hence the issue.

### Alternative approaches

We can generally solve this in a couple of different ways, checking one of the following:

1. The value of `apply_filters( 'woocommerce_admin_disabled' )`.
1. If the `admin-note` data store is available using `apply_filters( 'woocommerce_data_stores' )`.
1. If the attempt to add a note throws an exception.

I opted for the last approach as it does not require WCS&T to be aware of WC Admin's implementation details such as what exactly is disabled when `woocommerce_admin_disabled` is false, and data store names. We are already using the trait so checking if it throws an exception keeps WCS&T's dependencies to just this trait.

However, the *[Using the Admin Notes Inbox in WooCommerce](https://developer.woocommerce.com/2020/10/16/using-the-admin-notes-inbox-in-woocommerce/)* article and [these examples](https://woocommerce.github.io/woocommerce-admin/#/examples/activity-panel-inbox) interact with the data store directly by providing its name in calls to `WC_Data_Store::load()` so it might also be OK to replace the `try-catch` blocks with just one inside `can_add_wc_admin_notices()`.

### Related issue(s)



Fixes #2400.

### Steps to reproduce & screenshots/GIFs



**To reproduce the issue:**

1. Add the following code to the site:
    ```php
    add_filter( 'woocommerce_admin_disabled', '__return_true' );
    ```
1. The site throws a fatal error when trying to access wp-admin.

With the changes in this PR applied, a fatal error is not thrown.

**To verify no regressions happened:**

1. Set up a fresh site and install the plugin.
2. No notice is displayed.
3. Enable DHL live rates and refresh the schemas.
4. The notice appears. Click either of the buttons to dismiss it.
5. The notice is gone.
5. Reinstall the plugin and activate it with DHL live rates still enabled.
6. The notice appears immediately.

### Checklist
